HOWTO.md

```diff
@@ -0,0 +1,120 @@
+(i.e. one has to go from x-pencils to y-pencils to z-pencils to y-pencils to x-pencils)
+Correspondingly, the library provides 4 communication subroutines:
+```
+  call transpose_x_to_y(var_in,var_out)
+  call transpose_y_to_z(var_in,var_out)
+  call transpose_z_to_y(var_in,var_out)
+  call transpose_y_to_x(var_in,var_out)
+```
+The input array ``var_in`` and output array ``var_out`` are defined by the code using the library
+and contain distributed data for the correct pencil orientations.
+
+Note that the library is written using Fortran's generic interface so different data types are supported
+without user input. That means in and out above can be either real or complex arrays,
+the latter being useful for applications involving 3D Fast Fourier Transforms.
+Finally, before exit, applications should clean up the memory by:
+```
+  call decomp_2d_finalize
+```
+Detailed information about the decomposition API are available [here](https://2decomp-fft.github.io/pages/api_domain.html) 
+### Use of the FFT module
+To use the FFT programming interface, first of all, one additional Fortran module is needed:
+```
+  use decomp_2d_fft
+```
+Then one needs to initialise the FFT interface using
+```
+  call decomp_2d_fft_init(pencil, n1, n2, n3)
+```
+where ``pencil=PHYSICAL_IN_X`` or ``PHYSICAL_IN_Z`` and ``n1, n2, n3`` is an arbitrary problem size
+that can be different from :math:`nx\times ny\times nz`.
+For complex-to-complex (c2c) FFTs, the user interface is:
+```
+  call decomp_2d_fft_3d(input, output, direction)
+```
+where ``direction`` can be either ``DECOMP_2D_FFT_FORWARD == -1`` for forward transforms, 
+or ``DECOMP_2D_FFT_BACKWARD == 1`` for backward transforms.
+We recommend using the ``DECOMP_2D_FFT_XXX`` variables, rather than literal ``1`` or ``-1``,
+to avoid potential issues if these values change in future versions.
+The input array (``input``) and the output one (``output``) are both complex
+and have to be either a X-pencil/Z-pencil combination or vice-versa.
+The interface for real-to-complex and complex-to-real transform is
+```
+  call decomp_2d_fft_3d(input, output)
+```
+If the ``input`` data are real type a forward transform is assumed obtaining a complex ``output``.
+Similarly a backward FFT is computed if ``input`` is a complex array and ``output`` a real array.
+Finally, to release the memory used by the FFT interface:
+```
+  call decomp_2d_fft_finalize
+```
+Detailed information about the FFT API are available [here](https://2decomp-fft.github.io/pages/api_fft.html) 
+### Use of the IO module
+All the I/O functions have been packed in a Fortran module:
+```
+  use decomp_2d_io
+```
+To write a single three-dimensional array to a file
+```
+  call decomp_2d_write_one(ipencil,var,directory,filename,icoarse,io_name)
+```
+where ``ipencil`` describes how the data is distributed (valid values are: 1 for X-pencil; 2 for
+Y-pencil and 3 for Z-pencil); ``var`` is the reference to the data array, which can be either real or
```

Comment:
``var`` is the data array to be written on the disk,

Reply:
I agree with this, the slight confusion is when using ADIOS2, if you use temporary arrays then you need to flush, otherwise you will get the wrong data, e.g.
```
tmp = a
call decomp_2d_write_one(1, tmp, ".", "a", 0, io_name)
tmp = b
call decomp_2d_write_one(1, tmp, "."., "b", 0, io_name)
```
will write the contents of array `b` to both files.

Reply:
Modified and added a comment to take into account ADIOS2. However, what is the correct command to flush? Just closing the file? 
`call decomp_2d_end_io(io_name, ".")`
therefore the above code should be 
```
tmp = a
call decomp_2d_write_one(1, tmp, ".", "a", 0, io_name)
call decomp_2d_end_io(io_name, ".")
tmp = b
call decomp_2d_write_one(1, tmp, ".", "b", 0, io_name)
```
correct?

Reply:
this is the fill text i add

In case of ADIOS2 backend for IO operation please remember that variables registred for I/O needs to  
be flush as: 
```
tmp = a 
call decomp_2d_start_io(io_name, "out")
call decomp_2d_write_one(1, tmp, "out", "a", 0, io_name) 
call decomp_2d_end_io(io_name, "out") 
tmp = b 
call decomp_2d_start_io(io_name, "out")
call decomp_2d_write_one(1, tmp, "out", "b", 0, io_name)
call decomp_2d_end_io(io_name, "out") 
call decomp_2d_close_io(io_name, "out")
```
otherwise the content of `b` will be used for writing both `a` and `b`variables.

Reply:
The start/end will advance the step counter (i.e. time) in the file you want to do
```
call decomp_2d_start_io(io_name, "out") ! Begin timestep

tmp = a 
call decomp_2d_write_one(1, tmp, "out", "a", 0, io_name) 
tmp = b 
call decomp_2d_write_one(1, tmp, "out", "b", 0, io_name)

call decomp_2d_end_io(io_name, "out") ! End timestep

call decomp_2d_close_io(io_name, "out") ! This could be right at the end of the simulation during finalise

Reply:
Then what we have to do to flush the data to avoid that `a` and `b` are going to be the same? i cannot find anything in the examples.

Reply:
Sorry, difficulty with splitting conversation between here and Slack - I made a typo here
```
call decomp_2d_start_io(io_name, "out") ! Begin timestep

tmp = a 
call decomp_2d_write_one(1, tmp, "out", "a", 0, io_name, opt_deferred_writes=.false.) 
tmp = b 
call decomp_2d_write_one(1, tmp, "out", "b", 0, io_name, opt_deferred_writes=.false.)

call decomp_2d_end_io(io_name, "out") ! End timestep

call decomp_2d_close_io(io_name, "out") ! This could be right at the end of the simulation during finalise
```
you're right, I'll extend the examples to test this functionality.

Reply:
ok for the HOWTO I'll keep only a generic sentence and when we sort it we can update with an example or something.

Reply:
Ok this seems to be resolved with the new text and relarive explanation.